### ISSUE TITLE
[feature request]: checking for image size before import/bootstrap/other

### ISSUE THREAD

**Issue by ityonemo** [apptainer/singularity#646]
Do you think it's reasonable to have a minimum size check for the singularity container bootstrap definitions? Perhaps a %size field in the container that checks to see if the existing image has at a minimum enough space, and halts if the .img file being bootstrapped is large enough.

**Comment by vsoch** [comment]
hey @ityonemo ! I think this would be very reasonable / doable to implement, even without having the user specify a `%size` field.  Here are a few cases that come off the top of my head:

## Case 1: User is making a new image
This would be akin to doing a bootstrap or import into a newly created image. Given that we are adding docker layers to something, it is reasonable that, given that the layers dump is too big for the image that exists or is being generated, we could issue a warning. If we had create and bootstrap (or import) wrapped into one command (eg, @gmkurtzer this would have been like the docker pull I've been poking about) then the size of the image would be already available to compare against the size of the layers.  This would be fairly accurate given that the image is not yet generated, because we have confidence that most of the dump of content is going to fit in the empty image. So in summary:
 - given a new image with pull (bootstrap + create) we compare layers size to image size, stop if overflow
 - given a pre-existing empty image (bootstrap or import after create) we get size of image first, and compare to layers, stop if overflow

## Case 2: User is importing into existing image
This (I hope) is less likely done, because it would mean a less reproducible practice, but you can imagine importing layers into an already existing image. In this case, the most we can do is estimate the size of content in the image, estimate the size of the layers, and issue a warning (but probably not stop the process) if the imported stuffs is greater than the image size. The reason is because some of the dump of the layers might just override stuff already there, so the size wouldn't be extra added. So in summary:
- given a pre-existing non empty image, we can't say what % of the image files will be overwritten, so the most we can do is give a warning.

## Case 3: User is bootstrapping
This case is most similar to what you are describing - the user has some idea about the size of the content in his bootstrap spec (meaning downloads and installs that aren't coming from the docker or other bootstrap). In this case, the user could specify some kind of `%size` parameter to check, but to play devil's advocate, couldn't he/she make this same specification when the image is created? Or is the idea to preserve this knowledge with the bootstrap?

TLDR - I think this is a good feature to think about and possibly add, and implementation wise it is very doable. I added an endpoint to the Python internal API that will return the size for Docker layers, and this could be used by the calling function to do a check before any time intensive stuffs are done. I'll chat with @gmkurtzer about if/when/where would be reasonable to implement this!

**Comment by vsoch** [comment]
Just for linking to the docker pull idea: https://github.com/singularityware/singularity/issues/531

**Comment by FlorianPommerening** [comment]
This feature would be just what I was looking for. In my use-case, person A creates a bootstrap script (starting from a docker image and containing some `apt-get install` calls, so its hard to predict the size). Afterwards person B creates and bootstraps the container. So far, A has to tell B what size to use for the container or B has to try with different sizes. With the suggested `%size` field, this information could be part of the bootstrap script.

At this point it would also be nice to have `singularity pull BOOTSTRAP_SCRIPT NAME` as an alias for `singularity create --size [%size from BOOTSTRAP_SCRIPT] NAME && singularity bootstrap NAME BOOTSTRAP_SCRIPT`, which is probably what you had in mind with the link to #531.

**Comment by FlorianPommerening** [comment]
Issue #138 also seems closely related to this.

**Comment by dtrudg** [comment]
Closing this issue - the referenced PR was closed previously:

> I am closing this PR as the new singularity build command which is superseding singularity bootstrap no longer requires a container size. Thank you, and please send us more PRs!
